ooni/reporter.py

```diff
@@ -206,6 +208,56 @@ def createReport(self):
     def finish(self):
         self._stream.close()
 
+class NJSONReporter(OReporter):
+
+    """
+    report_destination:
+        the destination directory of the report
+
+    """
+
+    def __init__(self, test_details, report_filename):
+        self.report_path = report_filename
+        OReporter.__init__(self, test_details)
+
+    def _writeln(self, line):
+        self._write("%s\n" % line)
+
+    def _write(self, data):
+        if not self._stream:
+            raise errors.ReportNotCreated
+        if self._stream.closed:
+            raise errors.ReportAlreadyClosed
+        s = str(data)
+        assert isinstance(s, type(''))
+        self._stream.write(s)
+        untilConcludes(self._stream.flush)
+
+    def writeReportEntry(self, entry):
+        if isinstance(entry, Measurement):
+            e = deepcopy(entry.testInstance.report)
+        elif isinstance(entry, dict):
+            e = deepcopy(entry)
+        else:
+            raise Exception("Failed to serialise entry")
+        report_entry = {
+            'input': e.pop('input', None),
+            'id': str(uuid.uuid4()),
+            'test_start_time': e.pop('test_start_time', None),
+            'measurement_start_time': e.pop('measurement_start_time', None),
+            'test_runtime': e.pop('test_runtime', None),
+            'test_keys': e
+        }
+        report_entry.update(self.testDetails)
+        self._write(json.dumps(report_entry))
+        self._write("\n")
```

Comment:
On second thought my concern with using `_writeln` is that it's doing string concatenation that can be expensive if the string to be concatenated is fairly large.

I will change how `writeln` works to be better at this.